File: development/NET/samples/client/DAConsole/VB/MyDaSession.vb
Imports System
Imports Softing.OPCToolbox.Client
Imports Softing.OPCToolbox

Namespace DaConsole
	Public Class MyDaSession
		Inherits DaSession

#Region "Constructor"
		'-----------------

		Public Sub New(ByVal url As String)
			MyBase.New(url)
			AddHandler StateChangeCompleted, AddressOf HandleStateChanged
			AddHandler PerformStateTransitionCompleted, AddressOf HandlePerformStateTransition
			AddHandler ShutdownRequest, AddressOf HandleServerShutdown
			AddHandler ReadCompleted, AddressOf HandleSessionReadCompleted
			AddHandler WriteCompleted, AddressOf HandleSessionWriteCompleted
			AddHandler GetStatusCompleted, AddressOf HandleGetServerStatus
			AddHandler LogonCompleted, AddressOf HandleLogonCompleted
			AddHandler LogoffCompleted, AddressOf HandleLogoffCompleted
		End Sub

		'--
#End Region

#Region "Private Members"
		'---------------------

#Region "Private Attributes"
		'------------------------


		'--
#End Region

		'--
#End Region

#Region "Public Methods"
		'---------------------

		'--
#End Region

#Region "Public Properties"
		'-----------------------


		'--
#End Region

#Region "Handles"
		'---------------------

		Public Shared Sub HandleStateChanged(ByVal sender As ObjectSpaceElement, ByVal state As EnumObjectState)
			System.Console.WriteLine(sender.ToString() + " State Changed - " + state.ToString())
		End Sub


		Public Shared Sub HandlePerformStateTransition(ByVal sender As ObjectSpaceElement, ByVal executionContext As UInt32, ByVal result As Integer)
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine(sender.ToString() + " Performed state transition - " + executionContext.ToString())
			Else
				System.Console.WriteLine(sender.ToString() + "  Performed state transition failed! Result: " + result.ToString())
			End If
		End Sub


		Public Shared Function HandleServerShutdown(ByVal sender As ObjectSpaceElement, ByVal reason As String) As Boolean
			System.Console.WriteLine()
			System.Console.WriteLine(sender.ToString() + " Shutdown - " + reason)
			System.Console.WriteLine()
			Return True
		End Function


		Public Shared Sub HandleSessionReadCompleted(ByVal daSession As DaSession, ByVal executionContext As UInt32, ByVal itemIds As String(), ByVal itemPaths As String(), ByVal values As ValueQT(), ByVal results As Integer(), _
		ByVal result As Integer)
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine(daSession.ToString() + " asynchronous read succeeded! ")
				Dim i As Integer = 0
				While i < itemIds.Length
					If ResultCode.SUCCEEDED(results(i)) Then
						System.Console.WriteLine([String].Format("{0,-19} {1} {2,-50} ", itemIds(i).ToString(), "-", values(i).ToString()))
					Else
						System.Console.WriteLine(" Session asynchronous read failed for item " + " Item: " + itemIds(i).ToString() + " [ResultID: " + results(i).ToString() + " ]")
					End If
					System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
				End While
			Else
				System.Console.WriteLine(" Session asynchronous read failed! Result: " + result.ToString())
			End If
		End Sub


		Public Shared Sub HandleSessionWriteCompleted(ByVal daSession As DaSession, ByVal executionContext As UInt32, ByVal itemIds As String(), ByVal itemPaths As String(), ByVal values As ValueQT(), ByVal results As Integer(), _
		ByVal result As Integer)
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine(daSession.ToString() + " asynchronous write succeeded! ")
				Dim i As Integer = 0
				While i < itemIds.Length
					If ResultCode.SUCCEEDED(results(i)) Then
						System.Console.WriteLine([String].Format("{0,-19} {1} {2,-50} ", itemIds(i).ToString(), "-", values(i).ToString()))
					Else
						System.Console.WriteLine(" Session asynchronous write for item " + itemIds(i).ToString() + " [ResultID: " + results(i).ToString() + " ]")
					End If
					System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
				End While
			Else
				System.Console.WriteLine("Session asynchronous write failed! Result:" + result.ToString())
			End If
		End Sub


		Public Shared Sub HandleGetServerStatus(ByVal sender As ObjectSpaceElement, ByVal executionContext As UInt32, ByVal serverStatus As ServerStatus, ByVal result As Integer)
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine(sender.ToString())
				System.Console.WriteLine("Server Status")
				System.Console.WriteLine("  Vendor info: " + serverStatus.VendorInfo)
				System.Console.WriteLine("  Product version: " + serverStatus.ProductVersion)
				System.Console.WriteLine("  State: " + serverStatus.State.ToString())
				System.Console.WriteLine("  Start time: " + serverStatus.StartTime.ToString())
				System.Console.WriteLine("  Last update time: " + serverStatus.LastUpdateTime.ToString())
				System.Console.WriteLine("  Current time: " + serverStatus.CurrentTime.ToString())
				System.Console.WriteLine("  GroupCount: " + serverStatus.GroupCount.ToString())
				System.Console.WriteLine("  Bandwidth: " + serverStatus.Bandwidth.ToString())
				Dim i As Integer = 0
				While i < serverStatus.SupportedLcIds.Length
					System.Console.WriteLine("  Supported LCID: " + serverStatus.SupportedLcIds(i).ToString())
					System.Math.Max(System.Threading.Interlocked.Increment(i), i - 1)
				End While
				System.Console.WriteLine("  Status info: " + serverStatus.StatusInfo)
			Else
				System.Console.WriteLine("Asynchronous get server status failed! Result: " + result.ToString())
			End If
		End Sub

		Public Shared Sub HandleLogonCompleted(ByVal sender As ObjectSpaceElement, ByVal executionContext As UInt32, ByVal username As String, ByVal password As String, ByVal result As Integer)
			System.Console.Write("Logon for secure communication")
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine("- successfull logon for user:" + username)
			Else
				System.Console.WriteLine("- failed logon:" + result.ToString)
			End If
		End Sub

		Public Shared Sub HandleLogoffCompleted(ByVal sender As ObjectSpaceElement, ByVal executionContext As UInt32, ByVal result As Integer)
			System.Console.Write("Logoff for secure communication")
			If ResultCode.SUCCEEDED(result) Then
				System.Console.WriteLine("- successfull logoff")
			Else
				System.Console.WriteLine("- failed logoff:" + result.ToString)
			End If
		End Sub
		'--
#End Region

	End Class

End Namespace

File: development/NET/samples/client/DAConsole/CS/MyDaSession.cs
using System;
using Softing.OPCToolbox.Client;
using Softing.OPCToolbox;

namespace DaConsole
{
	public class MyDaSession : DaSession
	{

		#region Constructor
		//-----------------

		public MyDaSession(string url) : base(url)
		{
			StateChangeCompleted += new StateChangeEventHandler(HandleStateChanged);
			PerformStateTransitionCompleted += new PerformStateTransitionEventHandler(HandlePerformStateTransition);
			ShutdownRequest += new ShutdownEventHandler(HandleServerShutdown);
			ReadCompleted += new SessionReadEventHandler(HandleSessionReadCompleted);
			WriteCompleted += new SessionWriteEventHandler(HandleSessionWriteCompleted);
			GetStatusCompleted += new GetStatusEventHandler(HandleGetServerStatus);
			LogonCompleted += new LogonEventHandler(HandleLogonCompleted);
			LogoffCompleted += new LogoffEventHandler(HandleLogoffCompleted);
		}

		//--
		#endregion

		#region Private Members
		//---------------------

		#region Private Attributes
		//------------------------


		//--
		#endregion

		//--
		#endregion

		#region Public Methods
		//---------------------

		//--
		#endregion

		#region Public Properties
		//-----------------------


		//--
		#endregion

		#region Handles
		//---------------------

		public static void HandleStateChanged(ObjectSpaceElement sender, EnumObjectState state)
		{
			System.Console.WriteLine(sender + " State Changed - " + state);
		} // end HandleStateChanged

		public static void HandlePerformStateTransition(
			ObjectSpaceElement sender,
			uint executionContext,
			int result)
		{
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine(sender + " Performed state transition - " + executionContext);
			}
			else
			{
				System.Console.WriteLine(sender + "  Performed state transition failed! Result: " + result);
			}
		} // end HandlePerformStateTransition


		public static bool HandleServerShutdown(ObjectSpaceElement sender, string reason)
		{
			System.Console.WriteLine();
			System.Console.WriteLine(sender + " Shutdown - " + reason);
			System.Console.WriteLine();
			return true; // reconnection will be performed
		} // end HandleServerShutdown


		public static void HandleSessionReadCompleted(
			DaSession daSession,
			uint executionContext,
			string[] itemIds,
			string[] itemPaths,
			ValueQT[] values,
			int[] results,
			int result)
		{
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine(daSession + " asynchronous read succeeded! ");
				for (int i = 0; i < itemIds.Length; i++)
				{
					if (ResultCode.SUCCEEDED(results[i]))
					{
						System.Console.WriteLine(String.Format("{0,-19} {1} {2,-50} ", itemIds[i], "-", values[i].ToString()));
					}
					else
					{
						System.Console.WriteLine(" Session asynchronous read failed for item " + " Item: " + itemIds[i] + " [ResultID: " + results[i] + " ]");
					}
				}
			}
			else
			{
				System.Console.WriteLine(" Session asynchronous read failed! Result: " + result);
			}
		} // end HandleSessionReadCompleted


		public static void HandleSessionWriteCompleted(
			DaSession daSession,
			uint executionContext,
			string[] itemIds,
			string[] itemPaths,
			ValueQT[] values,
			int[] results,
			int result)
		{
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine(daSession + " asynchronous write succeeded! ");
				for (int i = 0; i < itemIds.Length; i++)
				{
					if (ResultCode.SUCCEEDED(results[i]))
					{
						System.Console.WriteLine(String.Format("{0,-19} {1} {2,-50} ", itemIds[i], "-", values[i].ToString()));
					}
					else
					{
						System.Console.WriteLine(" Session asynchronous write for item " + itemIds[i] + " [ResultID: " + results[i] + " ]");
					}
				}
			}
			else
			{
				System.Console.WriteLine("Session asynchronous write failed! Result:" + result);
			}
		}// end HandleSessionWriteCompleted


		public static void HandleGetServerStatus(
			ObjectSpaceElement sender,
			uint executionContext,
			ServerStatus serverStatus,
			int result)
		{
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine(sender);
				System.Console.WriteLine("Server Status");
				System.Console.WriteLine("	Vendor info: " + serverStatus.VendorInfo);
				System.Console.WriteLine("	Product version: " + serverStatus.ProductVersion);
				System.Console.WriteLine("	State: " + serverStatus.State);
				System.Console.WriteLine("	Start time: " + serverStatus.StartTime);
				System.Console.WriteLine("	Last update time: " + serverStatus.LastUpdateTime);
				System.Console.WriteLine("	Current time: " + serverStatus.CurrentTime);
				System.Console.WriteLine("	GroupCount: " + serverStatus.GroupCount);
				System.Console.WriteLine("	Bandwidth: " + serverStatus.Bandwidth);
				for (int i = 0; i < serverStatus.SupportedLcIds.Length; i++)
				{
					System.Console.WriteLine("	Supported LCID: " + serverStatus.SupportedLcIds[i]);
				}
				System.Console.WriteLine("	Status info: " + serverStatus.StatusInfo);
			}
			else
			{
				System.Console.WriteLine("Asynchronous get server status failed! Result: " + result);
			}
		} // end HandleGetServerStatus


		public static void HandleLogonCompleted(
			ObjectSpaceElement sender,
			uint executionContext,
			String UserName,
			String Password,
			int result)
		{
			System.Console.Write("\n Logon for secure communication");
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine("- successfully logon for user:" + UserName);
			}
			else
			{
				System.Console.WriteLine("- failed logon for user:" + UserName + "  " + result);
			}//end if ... else
		}//end handleLogonCompleted

		public static void HandleLogoffCompleted(
			ObjectSpaceElement sender,
			uint executionContext,
			int result)
		{
			System.Console.Write("\n Logon for secure communication");
			if (ResultCode.SUCCEEDED(result))
			{
				System.Console.WriteLine("- successfully logoff");
			}
			else
			{
				System.Console.WriteLine("- failed logoff" + "  " + result);
			}//end if ... else
		}//end handleLogoffCompleted
		 //--
		#endregion

	}   //	end class MyDaSession

}   //	end namespace
